listings/ch02-applications/constant_product_amm/src/constant_product_amm.cairo

```diff
@@ -0,0 +1,259 @@
+        self.token0.write(IERC20Dispatcher { contract_address: token0 });
+        self.token1.write(IERC20Dispatcher { contract_address: token1 });
+    }
+
+    #[generate_trait]
+    impl PrivateFunctions of PrivateFunctionsTrait {
+        fn _mint(ref self: ContractState, to: ContractAddress, amount: u256) {
+            self.balance_of.write(to, self.balance_of.read(to) + amount);
+            self.total_supply.write(self.total_supply.read() + amount);
+        }
+
+        fn _burn(ref self: ContractState, from: ContractAddress, amount: u256) {
+            self.balance_of.write(from, self.balance_of.read(from) - amount);
+            self.total_supply.write(self.total_supply.read() - amount);
+        }
+
+        fn _update(ref self: ContractState, reserve0: u256, reserve1: u256) {
+            self.reserve0.write(reserve0);
+            self.reserve1.write(reserve1);
+        }
+
+        #[inline(always)]
+        fn select_token(self: @ContractState, token: ContractAddress) -> bool {
+            assert(
+                token == self.token0.read().contract_address
+                    || token == self.token1.read().contract_address,
+                'invalid token'
+            );
+            token == self.token0.read().contract_address
+        }
+
+        #[inline(always)]
+        fn min(x: u256, y: u256) -> u256 {
+            if (x <= y) {
+                x
+            } else {
+                y
+            }
+        }
+    }
+
+    #[external(v0)]
+    impl ConstantProductAmm of super::IConstantProductAmm<ContractState> {
+        fn swap(ref self: ContractState, token_in: ContractAddress, amount_in: u256) -> u256 {
+            assert(amount_in > 0, 'amount in = 0');
+            let is_token0: bool = self.select_token(token_in);
+
+            let (token0, token1): (IERC20Dispatcher, IERC20Dispatcher) = (
+                self.token0.read(), self.token1.read()
+            );
+            let (reserve0, reserve1): (u256, u256) = (self.reserve0.read(), self.reserve1.read());
+            let (
+                token_in, token_out, reserve_in, reserve_out
+            ): (IERC20Dispatcher, IERC20Dispatcher, u256, u256) =
+                if (is_token0) {
+                (token0, token1, reserve0, reserve1)
+            } else {
+                (token1, token0, reserve1, reserve0)
+            };
+
+            let caller = get_caller_address();
+            let this = get_contract_address();
+            token_in.transfer_from(caller, this, amount_in);
+
+            // How much dy for dx?
+            // xy = k
+            // (x + dx)(y - dy) = k
+            // y - dy = k / (x + dx)
+            // y - k / (x + dx) = dy
+            // y - xy / (x + dx) = dy
+            // (yx + ydx - xy) / (x + dx) = dy
+            // ydx / (x + dx) = dy
+            // 0.3% fee
+
+            let amount_in_with_fee = (amount_in * 997) / 1000;
```

Comment:
what do you think about making the 0.3% a storage var initialised in the constructor
i don't think its critical nice to have

Reply:
It could be nice to have it as a variable. I was trying to reproduce the exact same contract as the [Solidity By Example one](https://solidity-by-example.org/defi/constant-product-amm/).